plugins/modules/instance.py

```diff
@@ -602,6 +640,54 @@ def _compare_param_to_device(
             device_param
         )
 
+    @staticmethod
+    def _normalize_local_interface(
+        local_interface: Dict[str, Any], remote_interface: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        """
+        Normalizes the given param interface to the remote interface
+        for direct comparison.
+        """
+        result = copy.deepcopy(local_interface)
+
+        # The IPv4 field will be implicitly populated if is not defined
+        if "ipv4" not in local_interface and "ipv4" in remote_interface:
+            result["ipv4"] = remote_interface.get("ipv4")
+
+        # Primary is only allowed for public and VPC purposes, so we
+        # should implicitly populate a default
+        if (
+            local_interface.get("purpose") in ("public", "vpc")
+            and "primary" not in local_interface
+        ):
+            result["primary"] = False
+
```

Comment:
Here's the fix for the test failure (`instance_config_vlan`) I mentioned earlier:

```suggestion

        # The primary field will not be returned for VLAN interfaces,
        # so we should drop it from the user-configured interface.
        if local_interface.get("purpose") == "vlan" and "primary" in result:
            primary = result.pop("primary")

            # Extra validation step to make sure users aren't trying to
            # set a VLAN as a primary interface.
            if primary:
                raise ValueError("VLAN interfaces cannot be primary interfaces")
```